File: data/readable_format/fortran/file1421.f90
module kind_types
  implicit none
  integer, parameter :: dp = selected_real_kind(15)
  integer, parameter :: qp = selected_real_kind(33)
  complex(dp), parameter :: cI_dp = (0._dp, 1._dp)
  complex(qp), parameter :: cI_qp = (0._qp, 1._qp)

contains

  function cmplx_dp(re, im)
    real(dp), intent(in) :: re
    real(dp), intent(in), optional :: im
    complex(dp) :: cmplx_dp
    if (present(im)) then
      cmplx_dp = complex(re, im)
    else
      cmplx_dp = complex(re, 0._dp)
    endif
  end function cmplx_dp

  function cmplx_qp(re, im)
    real(qp), intent(in) :: re
    real(qp), intent(in), optional :: im
    complex(qp) :: cmplx_qp
    if (present(im)) then
      cmplx_qp = complex(re, im)
    else
      cmplx_qp = complex(re, 0._qp)
    endif
  end function cmplx_qp

end module kind_types

program test_kind_types
  use kind_types
  implicit none
  complex(dp) :: result_dp
  complex(qp) :: result_qp

  ! Testing cmplx_dp
  result_dp = cmplx_dp(1.0_dp)
  print *, "cmplx_dp with only real part: ", result_dp

  result_dp = cmplx_dp(1.0_dp, 1.0_dp)
  print *, "cmplx_dp with real and imaginary parts: ", result_dp

  ! Testing cmplx_qp
  result_qp = cmplx_qp(1.0_qp)
  print *, "cmplx_qp with only real part: ", result_qp

  result_qp = cmplx_qp(1.0_qp, 1.0_qp)
  print *, "cmplx_qp with real and imaginary parts: ", result_qp
  
end program test_kind_types

File: data/readable_format/cpp/file1421.cpp
#include <gtest/gtest.h>
#include <complex>
#include <cmath>

// Define the kind_types class directly in this file
class kind_types {
public:
    static const int MaxParticles = 9;

    using intkind1 = int;
    using intkind2 = int;

    using dp = double;
    using qp = long double;

    static const std::complex<dp> cI_dp;
    static const std::complex<qp> cI_qp;

    static std::complex<dp> cmplx_dp(dp re, dp im = 0.0) {
        return std::complex<dp>(re, im);
    }

    static std::complex<qp> cmplx_qp(qp re, qp im = 0.0) {
        return std::complex<qp>(re, im);
    }
};

// Initialize the static constants
const std::complex<kind_types::dp> kind_types::cI_dp = std::complex<kind_types::dp>(0.0, 1.0);
const std::complex<kind_types::qp> kind_types::cI_qp = std::complex<kind_types::qp>(0.0, 1.0);

// Unit tests
TEST(kind_typesTest, cmplx_dp_with_imag) {
    std::complex<kind_types::dp> result = kind_types::cmplx_dp(1.0, 1.0);
    EXPECT_NEAR(result.real(), 1.0, 1e-10);
    EXPECT_NEAR(result.imag(), 1.0, 1e-10);
}

TEST(kind_typesTest, cmplx_dp_without_imag) {
    std::complex<kind_types::dp> result = kind_types::cmplx_dp(1.0);
    EXPECT_NEAR(result.real(), 1.0, 1e-10);
    EXPECT_NEAR(result.imag(), 0.0, 1e-10);
}

TEST(kind_typesTest, cmplx_qp_with_imag) {
    std::complex<kind_types::qp> result = kind_types::cmplx_qp(1.0L, 1.0L);
    EXPECT_NEAR(result.real(), 1.0L, 1e-10);
    EXPECT_NEAR(result.imag(), 1.0L, 1e-10);
}

TEST(kind_typesTest, cmplx_qp_without_imag) {
    std::complex<kind_types::qp> result = kind_types::cmplx_qp(1.0L);
    EXPECT_NEAR(result.real(), 1.0L, 1e-10);
    EXPECT_NEAR(result.imag(), 0.0L, 1e-10);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}